espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/Http/Router.java

```diff
@@ -1,113 +1,192 @@
-
-                // Convert route to a regex to test incoming URI against
-                // TODO: Cache these regexes instead of re-creating them every time
-                int index = 0;
-                for (String uriToken : testUri.split("/")) {
-                    if (uriToken.startsWith(":")) {
-                        testRegex = testRegex.concat("/[\\w\\W]*");
-                        wildcardIndices.put(index, uriToken.substring(1));
-                    } else if (!uriToken.equals("")) {
-                        testRegex = testRegex.concat("/" + uriToken);
-                    }
-                    index++;
-                }
-                testRegex += "$";
-
-                // If we have a match, parse the URI params and call that handler
-                if (uri.matches(testRegex)) {
-                    // TODO: Move this to a separate method 'parseUriParams'
-                    String[] uriTokens = uri.split("/");
-                    for (Map.Entry<Integer, String> wildcardIndexEntry : wildcardIndices.entrySet()) {
-                        int wildcardIndex = wildcardIndexEntry.getKey();
-                        uriParams.put(wildcardIndexEntry.getValue(), uriTokens[wildcardIndex]);
-                    }
-                    handler = routerMap.get(method).get(entry.getKey());
-                    break;
+                index++;
+            }
+            testRegex += "$";
+
+            // If we have a match, parse the URI params and call that handler
+            if (uri.matches(testRegex)) {
+                // TODO: Move this to a separate method 'parseUriParams'
+                String[] uriTokens = uri.split("/");
+                for (Map.Entry<Integer, String> wildcardIndexEntry : wildcardIndices.entrySet()) {
+                    int wildcardIndex = wildcardIndexEntry.getKey();
+                    uriParams.put(wildcardIndexEntry.getValue(), uriTokens[wildcardIndex]);
                 }
+                handler = routerMap.get(method).get(entry.getKey());
+                paramClass = paramClassMap.get(method).get(entry.getKey());
+                break;
             }
+        }
 
-            return handler.handle(session, uriParams);
-        } catch (Exception e) {
-            // TODO: Don't show internal error messages in production, only show them in dev
-            return new InternalErrorResponse(e.getMessage());
+        if (handler == null) {
+            return new ErrorResponse(AppiumStatus.UNKNOWN_COMMAND);
+        }
+
+        // Parse it to an Appium param
+        Gson gson = new Gson();
+        String postJson = parseBody(session);
+        AppiumParams appiumParams;
+        // TODO: Need to find a way to specify 'required' fields and throw exception when not provided
+        if (postJson == null) {
+            appiumParams = new AppiumParams();
+        } else {
+            appiumParams = (AppiumParams) paramClass.cast(gson.fromJson(postJson, paramClass));
```

Comment:
We can create Gson instance here, because it is not used anywhere else (or, probably, cache it on class level)